R/generic_import.R

```diff
@@ -0,0 +1,158 @@
+
+    #' @description
+    #' Tidy up open connections, rolling back any active transactions
+    close_connection = function() {
+      if (private$transaction_active) {
+        message("Rolling back active transaction")
+        self$rollback_transaction()
+      }
+      ## Connection will always be null on first call to reload
+      if (!is.null(private$con)) {
+        tryCatch(
+          DBI::dbDisconnect(private$con),
+          error = function(e) {
+            message("While disconnecting from db, ignored error:\n", e$message)
+          }
+        )
+      }
+    }
+  ),
+
+  public = list(
+    #' @field path Path to directory containing import object
+    path = NULL,
+
+    #' @description
+    #' Create Import object - shouldn't be called directly. Use
+    #' \code{\link{dettl}} to create the object
+    #' @param path Path to directory containing import object
+    #' @param db_name Database from dettl config to create import object for
+    #' @return A new `Import` object
+    initialize = function(path, db_name) {
+      self$path <- normalizePath(path, winslash = "/", mustWork = TRUE)
+      private$db_name <- db_name
+    },
+
+    #' @description
+    #' Reload the objects sources to refresh source code or repair a broken
+    #' Postgres connection.
+    reload = function() {
+      private$import_config <- read_config(self$path)
+      private$modify_in_transaction <- private$import_config$dettl$transaction
+      private$repo_config <- dettl_config(self$path)
+
+      private$db_name <- private$db_name %||%
+        get_default_type(private$repo_config)
+      private$close_connection()
+      private$con <- db_connect(private$db_name, self$path)
+
+      private$log_table <- db_get_log_table(private$db_name, self$path)
+      private$require_branch <-
+        private$repo_config$db[[private$db_name]]$require_branch
+
+      private$mode <- private$import_config$dettl$mode
+
+      private$confirm <- private$repo_config$db[[private$db_name]]$confirm
+    },
+
+    #' @description
+    #' Run checks that db can be modified, this checks that:
+    #' * If require_branch set in cfg that import is for that branch
+    #' * If confirm TRUE asks users to confirm action will modify db
+    #' * If git is dirty, checks that user has explicitly said that is okay
+    pre_modify_checks = function(dry_run, allow_dirty_git) {
+      if (!is.null(private$require_branch)) {
+        if (git_branch(self$path) != private$require_branch) {
+          stop(sprintf("This import can only be run from the '%s' branch",
+                       private$require_branch), call. = FALSE)
+        }
+      }
+      if (private$confirm) {
+        confirmed <- askYesNo(
+          sprintf(
+            "About to upload to database '%s' are you sure you want to proceed?",
+            private$db_name),
+          default = FALSE)
+        if (is.na(confirmed) || !confirmed) {
+          message("Not uploading to database.")
```

Comment:
Also I think could tidy this up, every other check throws an error but this one prints a message and returns. The return means the RImport class has a funny thing where it checks the result of the call to these checks and then returns itself if it is `FALSE` which feels a bit awkward. 

I thought about making this an error too but then that didn't feel good either because it is an expected user input. Any thoughts on nice way to make this a bit less awkward?